test-data/unit/merge.test

```diff
@@ -829,6 +829,110 @@ NameExpr:11: Overload(def (x: target.A<0>) -> target.A<0>, def (x: builtins.int<
 NameExpr:12: Overload(def (x: target.A<0>) -> target.A<0>, def (x: builtins.str<2>) -> builtins.str<2>)
 NameExpr:12: Overload(def (x: target.A<0>) -> target.A<0>, def (x: builtins.str<2>) -> builtins.str<2>)
 
+[case testOverloaded]
+import target
+[file target.py]
+from typing import overload
+class A: pass
+
+@overload
+def f(x: A) -> A: pass
+@overload
+def f(x: int) -> int: pass
+
+def f(x): pass
+[file target.py.next]
+from typing import overload
+
+class A: pass
+class B: pass
+
+@overload
+def f(x: A) -> B: pass
+@overload
+def f(x: str) -> str: pass
+
+def f(x): pass
+[out]
+MypyFile:1<0>(
```

Comment:
I hope you verified this file by hand -- I can't follow anything here in detail.

Reply:
Yeah I looked at it by hand. At some point I may update the output of these test cases, as currently it's super verbose.